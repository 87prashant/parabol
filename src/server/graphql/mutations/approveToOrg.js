import {GraphQLID, GraphQLNonNull, GraphQLString} from 'graphql';
import ApproveToOrgPayload from 'server/graphql/types/ApproveToOrgPayload';
import approveToOrg from 'server/safeMutations/approveToOrg';
import {getUserId, getUserOrgDoc, requireOrgLeader} from 'server/utils/authorization';
import publish from 'server/utils/publish';
import {INVITATION, NOTIFICATION, ORG_APPROVAL, ORGANIZATION} from 'universal/utils/constants';


export default {
  type: ApproveToOrgPayload,
  description: 'Approve an outsider to join the organization',
  args: {
    email: {
      type: new GraphQLNonNull(GraphQLString)
    },
    orgId: {
      type: new GraphQLNonNull(GraphQLID)
    }
  },
  async resolve(source, {email, orgId}, {authToken, dataLoader, socketId: mutatorId}) {
    const operationId = dataLoader.share();

    // AUTH
    const viewerId = getUserId(authToken);
    const userOrgDoc = await getUserOrgDoc(viewerId, orgId);
    requireOrgLeader(userOrgDoc);

    // RESOLUTION
    const subOptions = {mutatorId, operationId};
    const {
      removedRequestNotifications,
      removedOrgApprovals,
      newInvitations,
      inviteeApprovedNotifications,
      teamInviteNotifications
    } = await approveToOrg(email, orgId, viewerId);

    const invitationIds = newInvitations.map(({id}) => id);

    const data = {
      removedRequestNotifications,
      removedOrgApprovals,
      invitationIds,
      inviteeApprovedNotifications,
      teamInviteNotifications
    };

    // tell the other org leaders that the request has been closed
    publish(ORGANIZATION, orgId, ApproveToOrgPayload, data, subOptions);

    // tell the teammembers that the org approval has been replaced by an invitation
    const teamIds = Array.from(new Set(removedOrgApprovals.map(({teamId}) => teamId)));
    teamIds.forEach((teamId) => {
      const teamData = {...data, teamIdFilter: teamId};
      publish(ORG_APPROVAL, teamId, ApproveToOrgPayload, teamData, subOptions);
      publish(INVITATION, teamId, ApproveToOrgPayload, teamData, subOptions);
    });

    // tell the inviter that their invitee got approved and sent an invitation
    inviteeApprovedNotifications.forEach((notification) => {
      const {userIds: [inviterUserId]} = notification;
      publish(NOTIFICATION, inviterUserId, ApproveToOrgPayload, data, subOptions);
    });

    // tell the invitee that they've been invited
    if (teamInviteNotifications.length) {
      const [teamInvite] = teamInviteNotifications;
      const {userIds: [inviteeUserId]} = teamInvite;
      publish(NOTIFICATION, inviteeUserId, ApproveToOrgPayload, data, subOptions);
    }
    return data;
  }
};

