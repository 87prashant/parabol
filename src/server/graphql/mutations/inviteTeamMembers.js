import {GraphQLID, GraphQLList, GraphQLNonNull} from 'graphql';
import Invitee from 'server/graphql/types/Invitee';
import InviteTeamMembersPayload from 'server/graphql/types/InviteTeamMembersPayload';
import inviteTeamMembers from 'server/safeMutations/inviteTeamMembers';
import {getUserId, requireOrgLeaderOrTeamMember} from 'server/utils/authorization';
import publish from 'server/utils/publish';
import {TEAM} from 'universal/subscriptions/constants';
import {ADDED, NOTIFICATION, ORG_APPROVAL, REJOIN_TEAM, REMOVED, TEAM_MEMBER} from 'universal/utils/constants';
import fromTeamMemberId from 'universal/utils/relay/fromTeamMemberId';

export default {
  type: new GraphQLNonNull(InviteTeamMembersPayload),
  description: `If in the org,
     Send invitation emails to a list of email addresses, add them to the invitation table.
     Else, send a request to the org leader to get them approval and put them in the OrgApproval table.`,
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The id of the inviting team'
    },
    invitees: {
      type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(Invitee)))
    }
  },
  async resolve(source, {invitees, teamId}, {authToken, dataLoader, socketId: mutatorId}) {
    const operationId = dataLoader.share();
    // AUTH
    await requireOrgLeaderOrTeamMember(authToken, teamId);
    const viewerId = getUserId(authToken);

    // RESOLUTION
    const subOptions = {mutatorId, operationId};
    const {
      orgApprovalIds,
      reactivations,
      results,
      removedOrgApprovals,
      removedRequestNotifications,
      requestNotifications
    } = await inviteTeamMembers(invitees, teamId, viewerId, subOptions);
    const reactivatedTeamMemberIds = reactivations.map(({teamMemberId}) => teamMemberId);

    // HANDLE REACTIVATION
    reactivations.forEach(({notificationId, teamMemberId, preferredName}) => {
      // send a team member + temporary toast to the rest of the users
      const notification = {type: REJOIN_TEAM, teamId, preferredName};
      publish(TEAM_MEMBER, teamId, ADDED, {teamId, notification}, subOptions);
      // send a team + persisted notification to the reactivated team member
      const {userId} = fromTeamMemberId(teamMemberId);
      publish(TEAM, userId, ADDED, {teamId, notificationId}, subOptions);
    });

    // HANDLE NON-ORG LEADERS NEEDING APPROVAL
    orgApprovalIds.forEach((orgApprovalId) => {
      publish(ORG_APPROVAL, teamId, ADDED, {orgApprovalId}, subOptions);
    });
    requestNotifications.forEach((requestNotification) => {
      const {id: notificationId, userIds} = requestNotification;
      userIds.forEach((userId) => {
        // if the org leader triggers this it won't be in the mutation payload, so mutatorId is not passed in
        publish(NOTIFICATION, userId, ADDED, {notificationId}, {operationId});
      });
    });

    // HANDLE ORG LEADER INVITING A PENDING ORG APPROVAL INVITEE
    const removedOrgApprovalIds = removedOrgApprovals.map(({id}) => id);
    removedOrgApprovalIds.forEach((removedOrgApprovalId) => {
      publish(ORG_APPROVAL, teamId, REMOVED, {removedOrgApprovalId}, subOptions);
    });
    removedRequestNotifications.forEach((notification) => {
      const {userIds} = notification;
      userIds.forEach((userId) => {
        // if the org leader triggers this it won't be in the mutation payload, so mutatorId is not passed in
        publish(NOTIFICATION, userId, REMOVED, {notification}, {operationId});
      });
    });
    return {orgApprovalIds, reactivatedTeamMemberIds, removedOrgApprovalIds, results};
  }
};

