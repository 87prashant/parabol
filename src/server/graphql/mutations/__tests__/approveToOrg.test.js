import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import MockPubSub from 'server/__mocks__/MockPubSub';
import socket from 'server/__mocks__/socket';
import makeDataLoader from 'server/__tests__/setup/makeDataLoader';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import expectAsyncToThrow from 'server/__tests__/utils/expectAsyncToThrow';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import serializeGraphQLType from 'server/__tests__/utils/serializeGraphQLType';
import getRethink from 'server/database/rethinkDriver';
import * as hashInviteTokenKey from 'server/graphql/models/Invitation/inviteTeamMembers/hashInviteTokenKey';
import approveToOrg from 'server/graphql/mutations/approveToOrg';
import {REQUEST_NEW_USER} from 'universal/utils/constants';
import * as sendEmailPromise from 'server/email/sendEmail';

MockDate.set(__now);
console.error = jest.fn();

describe('approveToOrg', () => {
  test('for a 1-team approval, sends teamInvite, clears requestNewUser, sends inviteeApproved', async () => {
    // SETUP
    hashInviteTokenKey.default = jest.fn(() => Promise.resolve('HA$H'));
    sendEmailPromise.default = jest.fn(() => true);
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockPubSub = new MockPubSub();
    const mockDB = new MockDB();
    await mockDB.init()
      .user(1)
      .newNotification(undefined, {type: REQUEST_NEW_USER})
      .newUser({name: 'invitee', email: mockDB.context.notification.inviteeEmail})
      .newOrgApproval({email: mockDB.context.notification.inviteeEmail});
    const {notification} = mockDB.context;
    const approver = mockDB.db.user[0];
    const authToken = mockAuthToken(approver);
    const dataLoader = makeDataLoader(authToken);
    const orgId = mockDB.context.organization.id;
    const email = notification.inviteeEmail;

    // TEST
    const res = await approveToOrg.resolve(undefined, {email, orgId}, {authToken, dataLoader, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      notification: r.table('Notification').getAll(orgId, {index: 'orgId'}).orderBy('teamName', 'type'),
      orgApproval: r.table('OrgApproval').getAll(email, {index: 'email'}).orderBy('createdAt'),
      invitation: r.table('Invitation').getAll(email, {index: 'email'}).orderBy('teamId')
    }, dynamicSerializer);

    expect(db).toMatchSnapshot();
    expect(serializeGraphQLType(res, 'NotificationsClearedPayload', dynamicSerializer)).toMatchSnapshot();
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
  });

  test('for a 2-team approval with the same inviter, sends teamInvite, clears requestNewUser, sends inviteeApproved', async () => {
    // SETUP
    hashInviteTokenKey.default = jest.fn(() => Promise.resolve('HA$H'));
    sendEmailPromise.default = jest.fn(() => true);
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockPubSub = new MockPubSub();
    const mockDB = new MockDB();
    await mockDB.init()
      .user(1)
      .newNotification(undefined, {type: REQUEST_NEW_USER})
      .newUser({name: 'invitee', email: mockDB.context.notification.inviteeEmail})
      .newOrgApproval({email: mockDB.context.notification.inviteeEmail})
      .newTeam({name: 'team 2', orgId: mockDB.context.organization.id})
      .user(1)
      .newNotification(undefined, {type: REQUEST_NEW_USER, email: mockDB.context.notification.inviteeEmail})
      .newOrgApproval({email: mockDB.context.notification.inviteeEmail});

    const {notification} = mockDB.context;
    const approver = mockDB.db.user[0];
    const authToken = mockAuthToken(approver);
    const dataLoader = makeDataLoader(authToken);
    const orgId = mockDB.context.organization.id;
    const email = notification.inviteeEmail;

    // TEST
    const res = await approveToOrg.resolve(undefined, {email, orgId}, {authToken, dataLoader, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      notification: r.table('Notification').getAll(orgId, {index: 'orgId'}).orderBy('teamName', 'type'),
      orgApproval: r.table('OrgApproval').getAll(email, {index: 'email'}).orderBy('createdAt'),
      invitation: r.table('Invitation').getAll(email, {index: 'email'}).orderBy('teamId')
    }, dynamicSerializer);

    expect(db).toMatchSnapshot();
    expect(serializeGraphQLType(res, 'NotificationsClearedPayload', dynamicSerializer)).toMatchSnapshot();
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
  });

  test('for a 2-team approval with different inviters, sends teamInvite, clears requestNewUser, sends inviteeApproved', async () => {
    // SETUP
    hashInviteTokenKey.default = jest.fn(() => Promise.resolve('HA$H'));
    sendEmailPromise.default = jest.fn(() => true);
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockPubSub = new MockPubSub();
    const mockDB = new MockDB();
    await mockDB.init()
      .user(1)
      .newNotification(undefined, {type: REQUEST_NEW_USER})
      .newUser({name: 'invitee', email: mockDB.context.notification.inviteeEmail})
      .newOrgApproval({email: mockDB.context.notification.inviteeEmail})
      .newTeam({name: 'team 2', orgId: mockDB.context.organization.id})
      .user(2)
      .newNotification(undefined, {type: REQUEST_NEW_USER, email: mockDB.context.notification.inviteeEmail})
      .newOrgApproval({email: mockDB.context.notification.inviteeEmail});

    const {notification} = mockDB.context;
    const approver = mockDB.db.user[0];
    const authToken = mockAuthToken(approver);
    const dataLoader = makeDataLoader(authToken);
    const orgId = mockDB.context.organization.id;
    const email = notification.inviteeEmail;

    // TEST
    const res = await approveToOrg.resolve(undefined, {email, orgId}, {authToken, dataLoader, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      notification: r.table('Notification').getAll(orgId, {index: 'orgId'}).orderBy('teamName', 'type'),
      orgApproval: r.table('OrgApproval').getAll(email, {index: 'email'}).orderBy('createdAt'),
      invitation: r.table('Invitation').getAll(email, {index: 'email'}).orderBy('teamId')
    }, dynamicSerializer);

    expect(db).toMatchSnapshot();
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
    expect(serializeGraphQLType(res, 'NotificationsClearedPayload', dynamicSerializer)).toMatchSnapshot();
  });

  test('throws if the caller does not own the notification', async () => {
    // SETUP
    const mockDB = new MockDB();
    await mockDB.init()
      .newNotification(undefined, {type: REQUEST_NEW_USER});
    const {notification} = mockDB.context;
    const wrongUser = mockDB.db.user[1];
    const authToken = mockAuthToken(wrongUser);
    const dataLoader = makeDataLoader(authToken);

    // TEST
    const {id: dbNotificationId} = notification;
    await expectAsyncToThrow(
      approveToOrg.resolve(undefined, {dbNotificationId}, {authToken, dataLoader, socket}),
      [dbNotificationId, wrongUser.id]
    );
  });
});
