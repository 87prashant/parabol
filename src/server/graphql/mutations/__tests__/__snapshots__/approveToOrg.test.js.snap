// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`approveToOrg for a 1-team approval, sends teamInvite, clears requestNewUser, sends inviteeApproved 1`] = `
Object {
  "invitation": Array [
    Object {
      "createdAt": 2017-01-01T08:00:00.000Z,
      "email": "<<constant>>",
      "hashedToken": "HA$H",
      "id": "<<constant>>",
      "inviteCount": 1,
      "invitedBy": "<<constant>>",
      "teamId": "<<constant>>",
      "tokenExpiration": 2017-01-31T08:00:00.000Z,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
  "notification": Array [
    Object {
      "id": 0,
      "inviteeEmail": 2,
      "inviterName": "Vision",
      "orgId": 3,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamId": 4,
      "teamName": "Team America",
      "type": "INVITEE_APPROVED",
      "userIds": Array [
        5,
      ],
    },
    Object {
      "id": 1,
      "inviteeEmail": 2,
      "inviterName": "Vision",
      "orgId": 3,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamId": 4,
      "teamName": "Team America",
      "type": "TEAM_INVITE",
      "userIds": Array [
        6,
      ],
    },
  ],
  "orgApproval": Array [
    Object {
      "approvedBy": 8,
      "createdAt": 2017-01-01T07:00:00.000Z,
      "email": 2,
      "id": 7,
      "isActive": true,
      "orgId": 3,
      "status": "APPROVED",
      "teamId": 4,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;

exports[`approveToOrg for a 1-team approval, sends teamInvite, clears requestNewUser, sends inviteeApproved 2`] = `
Object {
  "deletedIds": Array [
    9,
  ],
}
`;

exports[`approveToOrg for a 1-team approval, sends teamInvite, clears requestNewUser, sends inviteeApproved 3`] = `
Object {
  "invitationAdded": Array [
    Object {
      "channelId": "<<constant>>",
      "message": Object {
        "invitationAdded": Object {
          "invitation": Object {
            "createdAt": 2017-01-01T08:00:00.000Z,
            "email": "<<constant>>",
            "hashedToken": "<<constant>>",
            "id": "<<constant>>",
            "inviteCount": 1,
            "invitedBy": "<<constant>>",
            "teamId": "<<constant>>",
            "tokenExpiration": 2017-01-31T08:00:00.000Z,
            "updatedAt": 2017-01-01T08:00:00.000Z,
          },
        },
        "operationId": 0,
      },
    },
  ],
  "notificationsAdded": Array [
    Object {
      "channelId": 5,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "id": 0,
              "inviteeEmail": 2,
              "inviterName": "Vision",
              "orgId": 3,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamId": 4,
              "teamName": "Team America",
              "type": "INVITEE_APPROVED",
              "userIds": Array [
                5,
              ],
            },
          ],
        },
        "operationId": 0,
      },
    },
    Object {
      "channelId": 6,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "id": 1,
              "inviteeEmail": 2,
              "inviterName": "Vision",
              "orgId": 3,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamId": 4,
              "teamName": "Team America",
              "type": "TEAM_INVITE",
              "userIds": Array [
                6,
              ],
            },
          ],
        },
        "operationId": 0,
      },
    },
  ],
  "notificationsCleared": Array [
    Object {
      "channelId": 8,
      "message": Object {
        "mutatorId": undefined,
        "notificationsCleared": Object {
          "deletedIds": Array [
            10,
          ],
        },
        "operationId": 0,
      },
    },
  ],
  "orgApprovalRemoved": Array [
    Object {
      "channelId": 4,
      "message": Object {
        "operationId": 0,
        "orgApprovalRemoved": Object {
          "orgApproval": Object {
            "approvedBy": 8,
            "createdAt": 2017-01-01T07:00:00.000Z,
            "email": 2,
            "id": 7,
            "isActive": true,
            "orgId": 3,
            "status": "APPROVED",
            "teamId": 4,
            "updatedAt": 2017-01-01T08:00:00.000Z,
          },
        },
      },
    },
  ],
}
`;

exports[`approveToOrg for a 2-team approval with different inviters, sends teamInvite, clears requestNewUser, sends inviteeApproved 1`] = `
Object {
  "invitation": Array [
    Object {
      "createdAt": 2017-01-01T08:00:00.000Z,
      "email": "<<constant>>",
      "hashedToken": "HA$H",
      "id": "<<constant>>",
      "inviteCount": 1,
      "invitedBy": "<<constant>>",
      "teamId": "<<constant>>",
      "tokenExpiration": 2017-01-31T08:00:00.000Z,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
    Object {
      "createdAt": 2017-01-01T08:00:00.000Z,
      "email": "<<constant>>",
      "hashedToken": "HA$H",
      "id": "<<constant>>",
      "inviteCount": 1,
      "invitedBy": "<<constant>>",
      "teamId": "<<constant>>",
      "tokenExpiration": 2017-01-31T08:00:00.000Z,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
  "notification": Array [
    Object {
      "id": 0,
      "inviteeEmail": 4,
      "inviterName": "Vision",
      "orgId": 5,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamId": 6,
      "teamName": "Team America",
      "type": "INVITEE_APPROVED",
      "userIds": Array [
        8,
      ],
    },
    Object {
      "id": 1,
      "inviteeEmail": 4,
      "inviterName": "Vision",
      "orgId": 5,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamId": 6,
      "teamName": "Team America",
      "type": "TEAM_INVITE",
      "userIds": Array [
        9,
      ],
    },
    Object {
      "id": 2,
      "inviteeEmail": 4,
      "inviterName": "War Machine",
      "orgId": 5,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamId": 7,
      "teamName": "team 2",
      "type": "INVITEE_APPROVED",
      "userIds": Array [
        10,
      ],
    },
    Object {
      "id": 3,
      "inviteeEmail": 4,
      "inviterName": "War Machine",
      "orgId": 5,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamId": 7,
      "teamName": "team 2",
      "type": "TEAM_INVITE",
      "userIds": Array [
        9,
      ],
    },
  ],
  "orgApproval": Array [
    Object {
      "approvedBy": 13,
      "createdAt": 2017-01-01T07:00:00.000Z,
      "email": 4,
      "id": 11,
      "isActive": true,
      "orgId": 5,
      "status": "APPROVED",
      "teamId": 6,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
    Object {
      "approvedBy": 13,
      "createdAt": 2017-01-01T07:00:00.001Z,
      "email": 4,
      "id": 12,
      "isActive": true,
      "orgId": 5,
      "status": "APPROVED",
      "teamId": 7,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;

exports[`approveToOrg for a 2-team approval with different inviters, sends teamInvite, clears requestNewUser, sends inviteeApproved 2`] = `
Object {
  "invitationAdded": Array [
    Object {
      "channelId": "<<constant>>",
      "message": Object {
        "invitationAdded": Object {
          "invitation": Object {
            "createdAt": 2017-01-01T08:00:00.000Z,
            "email": "<<constant>>",
            "hashedToken": "<<constant>>",
            "id": "<<constant>>",
            "inviteCount": 1,
            "invitedBy": "<<constant>>",
            "teamId": "<<constant>>",
            "tokenExpiration": 2017-01-31T08:00:00.000Z,
            "updatedAt": 2017-01-01T08:00:00.000Z,
          },
        },
        "operationId": 0,
      },
    },
    Object {
      "channelId": "<<constant>>",
      "message": Object {
        "invitationAdded": Object {
          "invitation": Object {
            "createdAt": 2017-01-01T08:00:00.000Z,
            "email": "<<constant>>",
            "hashedToken": "<<constant>>",
            "id": "<<constant>>",
            "inviteCount": 1,
            "invitedBy": "<<constant>>",
            "teamId": "<<constant>>",
            "tokenExpiration": 2017-01-31T08:00:00.000Z,
            "updatedAt": 2017-01-01T08:00:00.000Z,
          },
        },
        "operationId": 0,
      },
    },
  ],
  "notificationsAdded": Array [
    Object {
      "channelId": 8,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "id": 0,
              "inviteeEmail": 4,
              "inviterName": "Vision",
              "orgId": 5,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamId": 6,
              "teamName": "Team America",
              "type": "INVITEE_APPROVED",
              "userIds": Array [
                8,
              ],
            },
          ],
        },
        "operationId": 0,
      },
    },
    Object {
      "channelId": 9,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "id": 1,
              "inviteeEmail": 4,
              "inviterName": "Vision",
              "orgId": 5,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamId": 6,
              "teamName": "Team America",
              "type": "TEAM_INVITE",
              "userIds": Array [
                9,
              ],
            },
          ],
        },
        "operationId": 0,
      },
    },
    Object {
      "channelId": 10,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "id": 2,
              "inviteeEmail": 4,
              "inviterName": "War Machine",
              "orgId": 5,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamId": 7,
              "teamName": "team 2",
              "type": "INVITEE_APPROVED",
              "userIds": Array [
                10,
              ],
            },
          ],
        },
        "operationId": 0,
      },
    },
    Object {
      "channelId": 9,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "id": 3,
              "inviteeEmail": 4,
              "inviterName": "War Machine",
              "orgId": 5,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamId": 7,
              "teamName": "team 2",
              "type": "TEAM_INVITE",
              "userIds": Array [
                9,
              ],
            },
          ],
        },
        "operationId": 0,
      },
    },
  ],
  "notificationsCleared": Array [
    Object {
      "channelId": 13,
      "message": Object {
        "mutatorId": undefined,
        "notificationsCleared": Object {
          "deletedIds": Array [
            14,
            15,
          ],
        },
        "operationId": 0,
      },
    },
  ],
  "orgApprovalRemoved": Array [
    Object {
      "channelId": 6,
      "message": Object {
        "operationId": 0,
        "orgApprovalRemoved": Object {
          "orgApproval": Object {
            "approvedBy": 13,
            "createdAt": 2017-01-01T07:00:00.000Z,
            "email": 4,
            "id": 11,
            "isActive": true,
            "orgId": 5,
            "status": "APPROVED",
            "teamId": 6,
            "updatedAt": 2017-01-01T08:00:00.000Z,
          },
        },
      },
    },
    Object {
      "channelId": 7,
      "message": Object {
        "operationId": 0,
        "orgApprovalRemoved": Object {
          "orgApproval": Object {
            "approvedBy": 13,
            "createdAt": 2017-01-01T07:00:00.001Z,
            "email": 4,
            "id": 12,
            "isActive": true,
            "orgId": 5,
            "status": "APPROVED",
            "teamId": 7,
            "updatedAt": 2017-01-01T08:00:00.000Z,
          },
        },
      },
    },
  ],
}
`;

exports[`approveToOrg for a 2-team approval with different inviters, sends teamInvite, clears requestNewUser, sends inviteeApproved 3`] = `
Object {
  "deletedIds": Array [
    16,
    17,
  ],
}
`;

exports[`approveToOrg for a 2-team approval with the same inviter, sends teamInvite, clears requestNewUser, sends inviteeApproved 1`] = `
Object {
  "invitation": Array [
    Object {
      "createdAt": 2017-01-01T08:00:00.000Z,
      "email": "<<constant>>",
      "hashedToken": "HA$H",
      "id": "<<constant>>",
      "inviteCount": 1,
      "invitedBy": "<<constant>>",
      "teamId": "<<constant>>",
      "tokenExpiration": 2017-01-31T08:00:00.000Z,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
    Object {
      "createdAt": 2017-01-01T08:00:00.000Z,
      "email": "<<constant>>",
      "hashedToken": "HA$H",
      "id": "<<constant>>",
      "inviteCount": 1,
      "invitedBy": "<<constant>>",
      "teamId": "<<constant>>",
      "tokenExpiration": 2017-01-31T08:00:00.000Z,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
  "notification": Array [
    Object {
      "id": 0,
      "inviteeEmail": 4,
      "inviterName": "Vision",
      "orgId": 5,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamId": 6,
      "teamName": "Team America",
      "type": "INVITEE_APPROVED",
      "userIds": Array [
        8,
      ],
    },
    Object {
      "id": 1,
      "inviteeEmail": 4,
      "inviterName": "Vision",
      "orgId": 5,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamId": 6,
      "teamName": "Team America",
      "type": "TEAM_INVITE",
      "userIds": Array [
        9,
      ],
    },
    Object {
      "id": 2,
      "inviteeEmail": 4,
      "inviterName": "Vision",
      "orgId": 5,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamId": 7,
      "teamName": "team 2",
      "type": "INVITEE_APPROVED",
      "userIds": Array [
        8,
      ],
    },
    Object {
      "id": 3,
      "inviteeEmail": 4,
      "inviterName": "Vision",
      "orgId": 5,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamId": 7,
      "teamName": "team 2",
      "type": "TEAM_INVITE",
      "userIds": Array [
        9,
      ],
    },
  ],
  "orgApproval": Array [
    Object {
      "approvedBy": 12,
      "createdAt": 2017-01-01T07:00:00.000Z,
      "email": 4,
      "id": 10,
      "isActive": true,
      "orgId": 5,
      "status": "APPROVED",
      "teamId": 6,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
    Object {
      "approvedBy": 12,
      "createdAt": 2017-01-01T07:00:00.001Z,
      "email": 4,
      "id": 11,
      "isActive": true,
      "orgId": 5,
      "status": "APPROVED",
      "teamId": 7,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;

exports[`approveToOrg for a 2-team approval with the same inviter, sends teamInvite, clears requestNewUser, sends inviteeApproved 2`] = `
Object {
  "deletedIds": Array [
    13,
    14,
  ],
}
`;

exports[`approveToOrg for a 2-team approval with the same inviter, sends teamInvite, clears requestNewUser, sends inviteeApproved 3`] = `
Object {
  "invitationAdded": Array [
    Object {
      "channelId": "<<constant>>",
      "message": Object {
        "invitationAdded": Object {
          "invitation": Object {
            "createdAt": 2017-01-01T08:00:00.000Z,
            "email": "<<constant>>",
            "hashedToken": "<<constant>>",
            "id": "<<constant>>",
            "inviteCount": 1,
            "invitedBy": "<<constant>>",
            "teamId": "<<constant>>",
            "tokenExpiration": 2017-01-31T08:00:00.000Z,
            "updatedAt": 2017-01-01T08:00:00.000Z,
          },
        },
        "operationId": 0,
      },
    },
    Object {
      "channelId": "<<constant>>",
      "message": Object {
        "invitationAdded": Object {
          "invitation": Object {
            "createdAt": 2017-01-01T08:00:00.000Z,
            "email": "<<constant>>",
            "hashedToken": "<<constant>>",
            "id": "<<constant>>",
            "inviteCount": 1,
            "invitedBy": "<<constant>>",
            "teamId": "<<constant>>",
            "tokenExpiration": 2017-01-31T08:00:00.000Z,
            "updatedAt": 2017-01-01T08:00:00.000Z,
          },
        },
        "operationId": 0,
      },
    },
  ],
  "notificationsAdded": Array [
    Object {
      "channelId": 8,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "id": 0,
              "inviteeEmail": 4,
              "inviterName": "Vision",
              "orgId": 5,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamId": 6,
              "teamName": "Team America",
              "type": "INVITEE_APPROVED",
              "userIds": Array [
                8,
              ],
            },
          ],
        },
        "operationId": 0,
      },
    },
    Object {
      "channelId": 9,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "id": 1,
              "inviteeEmail": 4,
              "inviterName": "Vision",
              "orgId": 5,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamId": 6,
              "teamName": "Team America",
              "type": "TEAM_INVITE",
              "userIds": Array [
                9,
              ],
            },
          ],
        },
        "operationId": 0,
      },
    },
    Object {
      "channelId": 8,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "id": 2,
              "inviteeEmail": 4,
              "inviterName": "Vision",
              "orgId": 5,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamId": 7,
              "teamName": "team 2",
              "type": "INVITEE_APPROVED",
              "userIds": Array [
                8,
              ],
            },
          ],
        },
        "operationId": 0,
      },
    },
    Object {
      "channelId": 9,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "id": 3,
              "inviteeEmail": 4,
              "inviterName": "Vision",
              "orgId": 5,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamId": 7,
              "teamName": "team 2",
              "type": "TEAM_INVITE",
              "userIds": Array [
                9,
              ],
            },
          ],
        },
        "operationId": 0,
      },
    },
  ],
  "notificationsCleared": Array [
    Object {
      "channelId": 12,
      "message": Object {
        "mutatorId": undefined,
        "notificationsCleared": Object {
          "deletedIds": Array [
            15,
            16,
          ],
        },
        "operationId": 0,
      },
    },
  ],
  "orgApprovalRemoved": Array [
    Object {
      "channelId": 6,
      "message": Object {
        "operationId": 0,
        "orgApprovalRemoved": Object {
          "orgApproval": Object {
            "approvedBy": 12,
            "createdAt": 2017-01-01T07:00:00.000Z,
            "email": 4,
            "id": 10,
            "isActive": true,
            "orgId": 5,
            "status": "APPROVED",
            "teamId": 6,
            "updatedAt": 2017-01-01T08:00:00.000Z,
          },
        },
      },
    },
    Object {
      "channelId": 7,
      "message": Object {
        "operationId": 0,
        "orgApprovalRemoved": Object {
          "orgApproval": Object {
            "approvedBy": 12,
            "createdAt": 2017-01-01T07:00:00.001Z,
            "email": 4,
            "id": 11,
            "isActive": true,
            "orgId": 5,
            "status": "APPROVED",
            "teamId": 7,
            "updatedAt": 2017-01-01T08:00:00.000Z,
          },
        },
      },
    },
  ],
}
`;

exports[`approveToOrg throws if the caller does not own the notification 1`] = `"Unauthorized. User is not a Billing Leader for that organization"`;
